staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go

```diff
@@ -243,142 +285,348 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+					b := make([]byte, 5)
+					n, err = stream.Read(b)
+					if err != nil {
+						t.Fatalf("error reading from stream: %s", err)
+					}
+					if n != 5 {
+						t.Fatalf("expected to read 5 bytes, but actually read %d", n)
+					}
+					if e, a := "hello", string(b[0:n]); e != a {
+						t.Fatalf("expected '%s', got '%s'", e, a)
+					}
+
+					if testCase.proxyServerFunc != nil {
+						if !proxierCalled {
+							t.Fatal("expected to use a proxy but proxier in SpdyRoundTripper wasn't called")
+						}
+						if proxyCalledWithHost != serverURL.Host {
+							t.Fatalf("expected to see a call to the proxy for backend %q, got %q", serverURL.Host, proxyCalledWithHost)
+						}
+					}
+
+					var expectedProxyAuth string
+					if testCase.proxyAuth != nil {
+						encodedCredentials := base64.StdEncoding.EncodeToString([]byte(testCase.proxyAuth.String()))
+						expectedProxyAuth = "Basic " + encodedCredentials
+					}
+					if len(expectedProxyAuth) == 0 && proxyCalledWithAuth {
+						t.Fatalf("proxy authorization unexpected, got %q", proxyCalledWithAuthHeader)
+					}
+					if proxyCalledWithAuthHeader != expectedProxyAuth {
+						t.Fatalf("expected to see a call to the proxy with credentials %q, got %q", testCase.proxyAuth, proxyCalledWithAuthHeader)
+					}
+
+				})
+			}
+		})
+	}
+}
+
+type Interceptor struct {
+	Authorization       socks5.AuthContext
+	proxyCalledWithHost *string
+}
+
+func (i *Interceptor) GetAuthContext() (int, map[string]string) {
+	return int(i.Authorization.Method), i.Authorization.Payload
+}
+
+func (i *Interceptor) Rewrite(ctx context.Context, req *socks5.Request) (context.Context, *socks5.AddrSpec) {
+	*i.proxyCalledWithHost = req.DestAddr.Address()
+	i.Authorization = socks5.AuthContext(*req.AuthContext)
+	return ctx, req.DestAddr
+}
+
+// be sure to unset environment variable https_proxy (if exported) before testing, otherwise the testing will fail unexpectedly.
+func TestRoundTripSocks5AndNewConnection(t *testing.T) {
+	for _, redirect := range []bool{false, true} {
+		t.Run(fmt.Sprintf("redirect = %t", redirect), func(t *testing.T) {
+			socks5Server := func(creds *socks5.StaticCredentials, interceptor *Interceptor) *socks5.Server {
+				var conf *socks5.Config
+				if creds != nil {
+					authenticator := socks5.UserPassAuthenticator{Credentials: creds}
+					conf = &socks5.Config{
+						AuthMethods: []socks5.Authenticator{authenticator},
+						Rewriter:    interceptor,
+					}
+				} else {
+					conf = &socks5.Config{Rewriter: interceptor}
 				}
-				req, err := http.NewRequest("GET", server.URL, nil)
+
+				ts, err := socks5.New(conf)
 				if err != nil {
-					t.Fatalf("%s: Error creating request: %s", k, err)
+					t.Errorf("failed to create sock5 server: %v", err)
 				}
+				return ts
+			}
+
+			testCases := map[string]struct {
+				proxyAuth              *url.Userinfo
+				serverConnectionHeader string
+				serverStatusCode       int
+				shouldError            bool
+			}{
```

Comment:
I expected to see similar coverage to TestRoundTripAndNewConnection of an https backend with valid/invalid certificates, via the socks proxy ... from what I can see, tls backends via the socks proxy aren't being tested currently

Reply:
That's a fair point! I've cherry-picked most of the `TestRoundTripAndNewConnection` except that now they are going through the SOCKS5 proxy.